enhancements/20250407-cpu-model-selection.md

```diff
@@ -0,0 +1,471 @@
+
+```
+Architecture:                x86_64
+  CPU op-mode(s):            32-bit, 64-bit
+  Address sizes:             46 bits physical, 48 bits virtual
+Vendor ID:                   GenuineIntel
+  Model name:                12th Gen Intel(R) Core(TM) i9-12950HX
+    Flags:                   fpu vme de pse tsc msr pae ....
+```
+
+KubeVirt has a labeler to label the node with above information. Hence, you'll see below information on the node resource.
+
+```yaml
+kind: Node
+metadata:
+  name: my-node-1
+  labels:
+    host-model-cpu.node.kubevirt.io/SierraForest: "true"
+    cpu-model.node.kubevirt.io/IvyBridge: "true"
+    cpu-model.node.kubevirt.io/SandyBridge: "true"
+    cpu-model-migration.node.kubevirt.io/IvyBridge: "true"
+    cpu-model-migration.node.kubevirt.io/SandyBridge: "true"
+    cpu-feature.node.kubevirt.io/fpu: "true"
+    cpu-feature.node.kubevirt.io/vme: "true"
+```
+
+Harvester uses `host-model` as default CPU model. The migration process will be like:
+
+- Start a migration.
+- Fill out the nodeSelector in the new POD with
+  - A CPU model in `cpu-model-migration.node.kubevirt.io/SierraForest`
+  - Some features in `cpu-feature.node.kubevirt.io/xxx`.
+- Start the new POD.
+
+That being said, if the CPU Model is from a different generation, the migration might fail with default `host-model`. However, different CPU generations could still have some common CPU models. For example, let's use `IvyBridge` as a common cpu model in different CPU generations. That means if we use `IvyBridge` as the CPU model in the `virtualmachine` spec, this VM can be migrated to another node even though the nodes' CPU generations are different.
+
+## Proposal
+
+Due to multiple nodes, we can't show a big matrix for all the CPU models. Instead, we'll provide a new UI that allows users to select or input the CPU model.
+
+### User Stories
+
+#### Story 1
+
+I have multiple nodes that have a common CPU model called `Nehalem`. For some reason, I'll need to migrate my VMs to other nodes without manually shutting down. To ensure compatibility, I create my VMs with `Nehalem` as the CPU model.
+
+### User Experience In Detail
+
+If the cluster consists of mainly homogeneous nodes (including new nodes), users can define a universal common CPU model to ensure compatible migrations among the nodes.
+
+### API changes
+
+Backend should provide a API to let frontend geenrate a selection list.
+
+API: `GET /v1/harvester/node?link=getCpuMigrationCapabilities`
+
+```json
+{
+  "totalNodes": "<total number of ready nodes in the cluster>",
+  "globalModels": ["host-model", "host-passthrough", "one from kubevirt spec.configuration.cpuModel"],
+  "models": {
+    "<cpu-model-name>": {
+      "readyCount": "<number of ready nodes that support this CPU model>",
+      "migrationSafe": "<boolean: true if readyCount equals totalNodes, false otherwise>"
+    }
+  }
+}
+```
+
+**Field Descriptions:**
+- `totalNodes`: Total number of ready nodes in the cluster
```

Comment:
From the example cases below, I am not sure we need this and `migrationSafe`. Technically, migration is safe only if ready count > 1. Are there any situations where `migrationSafe == false` but `readyCount > 1`?

Reply:
> Are there any situations where migrationSafe == false but readyCount > 1?

No. It's just another way to show whether your node is migrated or not. In original thought, I considered migrationSafe as a status that you can migrate your VM to any nodes. Right now, the `readyCount` and `migrationSafe` are quite similar. 

But, I want to keep it flexible if we might need to show the how many nodes VM is able to be migrated to. So, I think we can keep it now. Also, It's not much effort to remove or add those fields. If we need to improve it, then I'll do it during implementation.